File: iot/iot/csl/js/src/shared/StorageObject.js
/**
 * Copyright (c) 2017, 2018, Oracle and/or its affiliates. All rights reserved.
 *
 * This software is dual-licensed to you under the MIT License (MIT) and
 * the Universal Permissive License (UPL). See the LICENSE file in the root
 * directory for license terms. You may choose either license, or both.
 *
 */

/**
 * StorageObject provides information about content in cloud storage.  For creation use
 * {@link iotcs.device.util.DirectlyConnectedDevice#createStorageObject}.
 *
 * @alias iotcs.StorageObject
 * @class
 * @extends iotcs.ExternalObject
 * @memberof iotcs
 *
 * @param {?string} uri - The full URI of the object in the Storage Cloud.
 * @param {?string} name - The name of the object used in the Storage Cloud.
 * @param {?string} type - The type of the object, if <code>null</code> then
 *        {@link iotcs.StorageObject.MIME_TYPE}.
 * @param {?string} encoding - The encoding of the object, or <code>null</code> if none.
 * @param {?Date} date - The last-modified date of the object.
 * @param {number} [length = -1] - The length of the object.
 */
lib.StorageObject = function (uri, name, type, encoding, date, length) {
    _optionalArg(uri, 'string');
    _optionalArg(name, 'string');
    _optionalArg(type, 'string');
    _optionalArg(encoding, 'string');
    _optionalArg(date, Date);
    _optionalArg(length, 'number');

    lib.ExternalObject.call(this, uri);

    var spec = {
        name: name || null,
        type: type || lib.StorageObject.MIME_TYPE,
        encoding: encoding || null,
        date: date || null,
        length: length || -1
    };
    var self = this;

    Object.defineProperties(this._.internal, {
        name: {
            value: spec.name,
            enumerable: true,
            writable: true
        },
        type: {
            value: spec.type,
            enumerable: true,
            writable: true
        },
        inputStream: {
            value: null,
            enumerable: true,
            writable: true
        },
        outputStream: {
            value: null,
            enumerable: true,
            writable: true
        },
        encoding: {
            value: spec.encoding,
            enumerable: true,
            writable: true
        },
        date: {
            value: spec.date,
            enumerable: true,
            writable: true
        },
        length: {
            value: spec.length,
            enumerable: true,
            writable: true
        },
        progress_state: {
            value: lib.StorageDispatcher.Progress.State.INITIATED,
            enumerable: true,
            writable: true
        }
    });

    Object.defineProperty(this._, 'dcd',{
        enumerable: false,
        configurable: false,
        writable: true,
        value: null
    });

    Object.defineProperty(this._, 'setDevice',{
        enumerable: false,
        configurable: false,
        writable: true,
        value: function (device) {
            if (device instanceof lib.device.util.DirectlyConnectedDevice) {
                self._.dcd = device;
            } else {
                lib.error("Invalid device type");
            }
        }
    });

    Object.defineProperty(this._, 'setMetadata',{
        enumerable: false,
        configurable: false,
        writable: false,
        value: function (date, length) {
            self._.internal.date = date;
            self._.internal.length = length;
        }
    });

    Object.defineProperty(this._, 'setURI',{
        enumerable: false,
        configurable: false,
        writable: false,
        value: function (uri) {
            self._.internal.uri = uri;
        }
    });

    Object.defineProperty(this._, 'setProgressState',{
        enumerable: false,
        configurable: false,
        writable: false,
        value: function (state) {
            self._.internal.progress_state = state;
        }
    });

    Object.defineProperty(this._, 'isCancelled',{
        enumerable: false,
        configurable: false,
        writable: false,
        value: function () {
            return (self._.internal.progress_state === lib.StorageDispatcher.Progress.State.CANCELLED);
        }
    });
};

lib.StorageObject.prototype = Object.create(lib.ExternalObject.prototype);
lib.StorageObject.constructor = lib.StorageObject;

/**
 * Set an input stream for content to be uploaded.  The implementation allows for either the input
 * stream to be set, or the output stream to be set, but not both.  If the input stream parameter is
 * not null, the output stream will be set to null.
 *
 * @function setInputStream
 * @memberof iotcs.StorageObject.prototype
 *
 * @param {stream.Readable} stream - A readable stream to which the content will be read.
 */
lib.StorageObject.prototype.setInputStream = function (stream) {
    _mandatoryArg(stream, require('stream').Readable);
    switch (this._.internal.progress_state) {
        case lib.StorageDispatcher.Progress.State.QUEUED:
        case lib.StorageDispatcher.Progress.State.IN_PROGRESS:
            lib.error("Can't set input stream during transfer process.");
            return;
        case lib.StorageDispatcher.Progress.State.COMPLETED:
            this._.internal.progress_state = lib.StorageDispatcher.Progress.INITIATED;
    }
    this._.internal.inputStream = stream;
    this._.internal.outputStream = null;
};

/**
 * Set an output stream for content to be downloaded.  The implementation allows for either the
 * output stream to be set, or the input stream to be set, but not both.  If the output stream
 * parameter is not null, the input stream will be set to null.
 *
 * @function setOutputStream
 * @memberof iotcs.StorageObject.prototype
 *
 * @param {stream.Writable} stream - A writable stream to which the content will be written.
 */
lib.StorageObject.prototype.setOutputStream = function (stream) {
    _mandatoryArg(stream, require('stream').Writable);
    switch (this._.internal.progress_state) {
        case lib.StorageDispatcher.Progress.State.QUEUED:
        case lib.StorageDispatcher.Progress.State.IN_PROGRESS:
            lib.error("Can't set output stream during transfer process.");
            return;
        case lib.StorageDispatcher.Progress.State.COMPLETED:
            this._.internal.progress_state = lib.StorageDispatcher.Progress.INITIATED;
    }
    this._.internal.outputStream = stream;
    this._.internal.inputStream = null;
};

/**
 * Get the the name of this object in the storage cloud.  This is name and path of the file that was
 * uploaded to the storage cloud.
 *
 * @function getName
 * @memberof iotcs.StorageObject.prototype
 *
 * @returns {string} The name of this object in the storage cloud.
 */
lib.StorageObject.prototype.getName = function () {
    return this._.internal.name;
};

/**
 * Get the mime-type of the content.
 *
 * @function getType
 * @memberof iotcs.StorageObject.prototype
 * @see {@link http://www.iana.org/assignments/media-types/media-types.xhtml|IANA Media Types}
 *
 * @returns {string} the mime-type of the content.
 */
lib.StorageObject.prototype.getType = function () {
    return this._.internal.type;
};

/**
 * Get the date and time the content was created or last modified in cloud storage.
 *
 * @function getDate
 * @memberof iotcs.StorageObject.prototype

 * @returns {?Date} The date the content was last modified in cloud storage, or <code>null</code> if
 *          the content has not been uploaded.
 */
lib.StorageObject.prototype.getDate = function () {
    return this._.internal.date;
};

/**
 * Get the length of the content in bytes.  This is the number of bytes required to upload or
 * download the content.
 *
 * @function getLength
 * @memberof iotcs.StorageObject.prototype
 *
 * @returns {number} The length of the content in bytes, or <code>-1</code> if unknown.
 */
lib.StorageObject.prototype.getLength = function () {
    return this._.internal.length;
};

/**
 * Get the compression scheme of the content.
 *
 * @function getEncoding
 * @memberof iotcs.StorageObject.prototype
 *
 * @returns {?string} The compression scheme of the content, or <code>null</code> if the content is
 *          not compressed.
 */
lib.StorageObject.prototype.getEncoding = function () {
    return this._.internal.encoding;
};

/**
 * Get the URI value.
 *
 * @function getURI
 * @memberof iotcs.StorageObject.prototype
 *
 * @returns {?string} The URI, or <code>null</code> if unknown.
 */
lib.StorageObject.prototype.getURI = function () {
    return this._.internal.uri;
};

/**
 * Get the input file path when uploading content.
 *
 * @function getInputStream
 * @memberof iotcs.StorageObject.prototype
 *
 * @returns {?stream.Readable} The input stream, or <code>null</code> if not set.
 */
lib.StorageObject.prototype.getInputStream = function () {
    return this._.internal.inputStream;
};

/**
 * Get the output file path when downloading content.
 *
 * @function getOutputStream
 * @memberof iotcs.StorageObject.prototype
 *
 * @returns {?stream.Writable} The output stream, or <code>null</code> if not set.
 */
lib.StorageObject.prototype.getOutputStream = function () {
    return this._.internal.outputStream;
};

/**
 * Synchronize content with the Storage Cloud Service.
 *
 * @function sync
 * @memberof iotcs.StorageObject.prototype
 *
 * @param {function(storage, error)} callback - The callback function.
 */
lib.StorageObject.prototype.sync = function (callback) {
    _mandatoryArg(callback, 'function');
    this._.dcd._.sync_storage(this, callback, callback);
};

/**
 * @constant MIME_TYPE
 * @memberof iotcs.StorageObject
 * @type {string}
 * @default "application/octet-stream"
 */
Object.defineProperty(lib.StorageObject, 'MIME_TYPE',{
    enumerable: false,
    configurable: false,
    writable: false,
    value: "application/octet-stream"
});

File: iot/java/iotcs/csl/javase/src/main/java/com/oracle/iot/client/StorageObject.java
/*
 * Copyright (c) 2017, Oracle and/or its affiliates.  All rights reserved.
 *
 * This software is dual-licensed to you under the MIT License (MIT) and
 * the Universal Permissive License (UPL).  See the LICENSE file in the root
 * directory for license terms.  You may choose either license, or both.
 */

package com.oracle.iot.client;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.security.GeneralSecurityException;

/**
 * StorageObject provides information about content in cloud storage.
 */
public abstract class StorageObject extends ExternalObject {

    /**
     * Get the the name of this object in the storage cloud. This is name and path of the file
     * that was uploaded to the storage cloud.
     * @return the name of this object in the storage cloud
     */
    public String getName() {
        return name;
    }

    /**
     * Get the mime-type of the content. See
     * <a href="http://www.iana.org/assignments/media-types/media-types.xhtml">IANA Media Types</a>.
     *
     * @return The mime-type of the content
     */
    public String getType() {
        return type;
    }

    /**
     * Get the compression scheme of the content.
     *
     * @return the compression scheme of the content, or {@code null} if the content is not compressed
     */
    public String getEncoding() {
        return encoding;
    }

    /**
     * Get the length of the content in bytes. This is the number of bytes required to upload or download
     * the content.
     *
     * @return The length of the content, or {@code -1} if unknown
     */
    public long getLength() {
        return length;
    }

    /**
     * Get the date and time the content was created or last modified in cloud 
     * storage. This may be {@code null} if the content has not been uploaded.
     * The date and time stamp format is ISO 8601.
     * @return The date the content was last modified in cloud storage, or {@code null} if the
     * content has not been uploaded.
     */
    public String getDate() {
        return date;
    }

    /**
     * Called by the library to obtain the input stream when uploading content
     * @return The input stream for the content being uploaded
     */
    public InputStream getInputStream() {
        return inputStream;
    }

    /**
     * Set an input stream for content to be uploaded. The implementation
     * allows for either the input stream to be set, or the output stream to be set, but not both.
     * If the {@code inputStream} parameter is not null, the output stream will be set to null.
     * @param inputStream A stream from which the content will be read
     * @throws IllegalStateException if a transfer is currently in progress
     */
    public void setInputStream(InputStream inputStream) {
        this.inputStream = inputStream;
        if (inputStream != null) {
            this.outputStream = null;
        }
    }

    /**
     * Called by the library to obtain the output stream when downloading content
     * @return The output stream for the content being downloaded
     */
    public OutputStream getOutputStream(){
        return outputStream;
    }

    /**
     * Set an output stream for content to be downloaded. The implementation
     * allows for either the input stream to be set, or the output stream to be set, but not both.
     * If the {@code outputStream} parameter is not {@code null}, the input stream will be set to
     * {@code null}.
     * @param outputStream A stream to which the content will be written
     * @throws IllegalStateException if a transfer is currently in progress
     */
    public void setOutputStream(OutputStream outputStream) {
        this.outputStream = outputStream;
        if (outputStream != null) {
            this.inputStream = null;
        }
    }

    /**
     * Synchronize content with the Storage Cloud Service.
     *
     * @throws IOException if there is an {@code IOException} raised by the runtime,
     * or a failure reported by the storage cloud
     * @throws GeneralSecurityException if there is an exception establishing a secure connection to the storage cloud
     * @throws IllegalArgumentException if the {@code StorageObject} does not return an {@code InputStream} or {@code OutputStream}
     */
    public abstract void sync() throws IOException, GeneralSecurityException;

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if(obj == null) return false;
        if (this.getClass() != obj.getClass()) return false;
        StorageObject other = (StorageObject)obj;
        if (!this.getURI().equals(other.getURI())) return false;
        if (!this.getName().equals(other.getName())) return false;
        if (!this.getType().equals(other.getType())) return false;
        return true;
    }

    @Override
    public int hashCode() {
        int hash = 5;
        hash = 61 * hash + this.name.hashCode();
        hash = 61 * hash + this.type.hashCode();
        hash = 61 * hash + this.getURI().hashCode();
        return hash;
    }
    
    /**
     * Used internally by the client library to update fields after upload. 
     * Not intended for general use.
     * @param date the last-modified date for the object in the Storage Cloud
     * @param length the length for the object in the Storage Cloud
     */
    protected void setMetadata(String date, long length) {
        this.date = date;
        this.length = length;
    }
    
    /**
     * Create a {@code StorageObject}.
     * @param uri the full URL of the object in the Storage Cloud
     * @param name name of the object used in the Storage Cloud
     * @param type type of the object, , or {@code null} if none
     * @param encoding encoding of the object, or {@code null} if none
     * @param date last-modified date of the object
     * @param length length of the object
     */
    protected StorageObject(String uri,
                            String name,
                            String type,
                            String encoding,
                            String date,
                            long length) {
        super(uri);
        this.name = name;
        this.type = type != null ? type : "application/octet-stream";
        this.encoding = encoding;
        this.date = date;
        this.length = length;
    }

    private final String name;
    private final String type;
    private final String encoding;
    private String date;
    private long length;
    private InputStream inputStream;
    private OutputStream outputStream;
}
